### PR TITLE
python binding for viz

### PR DESCRIPTION
**Pull request by LaurentBerger**
relative to https://github.com/opencv/opencv/issues/19490

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by LaurentBerger** [comment]
@alalek I don't know how to slove that https://pullrequest.opencv.org/buildbot/builders/precommit-contrib_linux64/builds/14900/steps/test_python3/logs/stdio

**Comment by alalek** [comment]
This is weird behavior from VTK. There is [call](https://github.com/Kitware/VTK/blob/d4e98d54de329c637f6737d36d8832ce0538f3ac/Rendering/OpenGL2/vtkXOpenGLRenderWindow.cxx#L465) of `abort()` if there is no display in environment.
As C++ library VTK should use C++ exceptions instead of `abort()`.

I can try to enable virtual display (xfvb) for tests, but it is bad practice to crash tests in non-supported environment. Perhaps we should check for DISPLAY environment variable on non-Windows (Linux) and skip such tests.
Or even in C++ code: check and throw exception (to survive from `abort()` call from VTK).

**Comment by LaurentBerger** [comment]
Is it possible to have a review?

File: modules/viz/include/opencv2/viz/types.hpp [hunk 1]
```diff
@@ -115,27 +114,87 @@ namespace cv
             static Color not_set();
         };
 
+        struct CV_EXPORTS_W_SIMPLE PyColor
```
Comment (alalek) [inline modules/viz/include/opencv2/viz/types.hpp:117]: We should not have Python-specific classes in public C++ header.

Please move this to `viz/misc/python/python_bridge.hpp` (see gapi module for example)

File: modules/viz/include/opencv2/viz/widgets.hpp [hunk 2]
```diff
@@ -55,6 +55,94 @@ namespace cv
...
+            WIDGET_WPaintedCloud,
+            WIDGET_WCloudCollection,
+            WIDGET_WCloudNormals,
+            WIDGET_WMesh,
+            WIDGET_WWidgetMerger
+        };
+/**
+PyAffine3 is an overloaded structure, provided for convenience.
+*/
+        struct CV_EXPORTS_W_SIMPLE PyAffine3 {
```
Comment (alalek) [inline modules/viz/include/opencv2/viz/widgets.hpp:88]: > PyAffine3

need to move out from public C++ headers

---

and similar classes below

File: modules/viz/src/viz3d.cpp [hunk 3]
```diff
@@ -44,8 +44,15 @@
 //M*/
 
 #include "precomp.hpp"
+#include <opencv2/core/utils/logger.hpp>
 
-cv::viz::Viz3d::Viz3d(const String& window_name) : impl_(0) { create(window_name); }
+cv::Ptr<cv::viz::Viz3d> cv::viz::Viz3d::Viz3d::create(const String& window_name)
+{
+    cv::utils::logging::setLogLevel(cv::utils::logging::LOG_LEVEL_WARNING);
```
Comment (alalek) [inline modules/viz/src/viz3d.cpp:51]: > `cv::utils::logging::setLogLevel(cv::utils::logging::LOG_LEVEL_WARNING);`

should no be called from OpenCV code. Remove that.

File: modules/viz/src/vizcore.cpp [hunk 4]
```diff
@@ -48,12 +48,19 @@
 cv::Affine3d cv::viz::makeTransformToGlobal(const Vec3d& axis_x, const Vec3d& axis_y, const Vec3d& axis_z, const Vec3d& origin)
 {
     Affine3d::Mat3 R(axis_x[0], axis_y[0], axis_z[0],
-                     axis_x[1], axis_y[1], axis_z[1],
-                     axis_x[2], axis_y[2], axis_z[2]);
+        axis_x[1], axis_y[1], axis_z[1],
+        axis_x[2], axis_y[2], axis_z[2]);
```
Comment (alalek) [inline modules/viz/src/vizcore.cpp:52]: Please avoid unnecessary changes.

File: modules/viz/src/clouds.cpp [hunk 5]
```diff
@@ -44,7 +44,7 @@
 //M*/
 
 #include "precomp.hpp"
-
+#include <iostream>
```
Comment (alalek) [inline modules/viz/src/clouds.cpp:47]: > `#include <iostream>`

remove that.

File: modules/viz/samples/viz_sample_03.py [hunk 6]
```diff
@@ -0,0 +1,60 @@
+#exec(open(r'G:\Lib\opencv_contrib\modules\viz\samples\viz_sample_03.py').read())
```
Comment (alalek) [inline modules/viz/samples/viz_sample_03.py:1]: No absolute paths.

File: modules/viz/samples/viz_sample_03.py [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import os
+os.add_dll_directory(r'G:\Lib\install\opencv\x64\vc15\bin')
+os.add_dll_directory(r'G:\Lib\install\vtk\bin')
+os.add_dll_directory(r'G:\Lib\install\ceres-solver\bin')
+os.add_dll_directory(r'G:\Lib\install\glog\bin')
+os.add_dll_directory(r'F:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.2\bin')
+os.add_dll_directory(r'F:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v11.2')
```
Comment (alalek) [inline modules/viz/samples/viz_sample_03.py:8]: It is for Python 3.8+ on Windows only.
There is no good solution for that from Python team.

There is some attempt to handle OpenCV paths through `<build_dir>/python_loader`.

Anyway such hacks should be removed from the source code.

File: modules/viz/samples/viz_sample_03.py [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    pose[0, 3] = 1
+    pose[1, 3] = 1
+    pose[1, 3] = 1
+    pose[0:3, 0] = v[0]
+    pose[0:3, 1] = v[1]
+    pose[0:3, 2] = v[2]
+    return pose
+
+def load_bunny():
+    with open(r'G:\Lib\opencv_contrib\modules\viz\samples\data\bunny.ply', 'r') as f:
```
Comment (alalek) [inline modules/viz/samples/viz_sample_03.py:24]: no absolute paths

File: modules/viz/misc/python/test/test_simple.py [hunk 9]
```diff
@@ -0,0 +1,436 @@
+#exec(open(r'G:\Lib\opencv_contrib\modules\viz\samples\viz_sample_03.py').read())
+import cv2 as cv
+import numpy as np
```
Comment (alalek) [inline modules/viz/misc/python/test/test_simple.py:3]: numpy should go first

File: modules/viz/include/opencv2/viz/widgets.hpp [hunk 10]
```diff
@@ -842,7 +930,833 @@ namespace cv
...
+        };
+
+        /** @brief This 3D Widget represents normals of a point cloud. :
+        */
+        struct CV_EXPORTS_W_SIMPLE PyWCloudNormals
+        {
+        public:
+			PyWCloudNormals()
+			{
+			}
```
Comment (alalek) [inline modules/viz/include/opencv2/viz/widgets.hpp:1717]: no tabs please

File: modules/viz/misc/python/test/test_simple.py [hunk 11]
```diff
@@ -0,0 +1,436 @@
...
+    cloud_pose_global = transform.product(cloud_pose)
+    myWindow.showWidget("CPW_FRUSTUM", cv.viz_PyWCameraPosition((0.889484, 0.523599)), camera_pose)
+    if not camera_pov:
+        myWindow.showWidget("CPW", cv.viz_PyWCameraPosition(0.5), camera_pose)
+    myWindow.showWidget("dragon", cloud_widget, cloud_pose_global)
+    if camera_pov:
+        myWindow.setViewerPosePy(camera_pose)
+
+class viz_test(NewOpenCVTests):
+
```
Comment (alalek) [inline modules/viz/misc/python/test/test_simple.py:43]: Lets skip VIZ tests by default (because they are UI tests):

```
    def setUp(self):
        super(viz_test, self).setUp()
        if not bool(os.environ.get('OPENCV_PYTEST_RUN_VIZ', False)):
            self.skipTest("Use OPENCV_PYTEST_RUN_VIZ=1 to enable VIZ UI tests")
```

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: modules/viz/include/opencv2/viz/types.hpp (hunk 1 above)
Comment (LaurentBerger) [reply modules/viz/include/opencv2/viz/types.hpp:117]: I need PyColor viz3d.hpp

 If I move PyColor defintion in python_bridge.hpp then it will generate error

File: modules/viz/misc/python/test/test_simple.py [hunk 12]
```diff
@@ -0,0 +1,440 @@
+import os
```
Comment (alalek) [inline modules/viz/misc/python/test/test_simple.py:1]: > modules/viz/misc/python/test/test_simple.py

Please rename to `test_viz_simple.py` (similar to python tests of other modules).

File: modules/viz/misc/python/test/test_simple.py [hunk 13]
```diff
@@ -0,0 +1,440 @@
...
+        myWindow.setViewerPosePy(camera_pose)
+
+class viz_test(NewOpenCVTests):
+    def setUp(self):
+        super(viz_test, self).setUp()
+        if not bool(os.environ.get('OPENCV_PYTEST_RUN_VIZ', False)):
+            self.skipTest("Use OPENCV_PYTEST_RUN_VIZ=1 to enable VIZ UI tests")
+
+    def test_viz_tutorial3_global_view(self):
+        tutorial3(False, self.find_file("viz\dragon.ply"))
```
Comment (alalek) [inline modules/viz/misc/python/test/test_simple.py:49]: > `viz\dragon.ply`

Use forward slash (/) instead in this file.

**Review by alalek — COMMENTED** [review]
Well done!

I will take a look on moving of "PyColor" and other classes problem a bit later.

BTW, Did you try to run it with enabled "Debug" mode (Debug assertions)? I see some failures:

```
$ OPENCV_PYTHON_DEBUG=1 OPENCV_PYTEST_RUN_VIZ=1 OPENCV_PYTEST_FILTER=test_simple* ./setup_vars.sh python3 <opencv>/modules/python/test/test.py -v --repo <opencv>

...

======================================================================
ERROR: test_viz_show_camera_positions (test_simple.viz_test)
----------------------------------------------------------------------
cv2.error: OpenCV(4.5.2-pre) /home/alalek/projects/opencv/dev/modules/core/include/opencv2/core/mat.inl.hpp:872: error: (-215:Assertion failed) (unsigned)(i1 * DataType<_Tp>::channels) < (unsigned)(size.p[1] * channels()) in function 'at'


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/alalek/projects/opencv/contrib/modules/viz/misc/python/test/test_simple.py", line 408, in test_viz_show_camera_positions
    viz.showWidget("pos2", cv.viz_PyWCameraPosition((0.78, 0.78), lena, 2.2, cv.viz_PyColor().green()), poses[0])
SystemError: <class 'cv2.viz_PyWCameraPosition'> returned a result with an error set

======================================================================
ERROR: test_viz_show_simple_widgets (test_simple.viz_test)
----------------------------------------------------------------------
cv2.error: OpenCV(4.5.2-pre) /home/alalek/projects/opencv/dev/modules/core/include/opencv2/core/mat.inl.hpp:872: error: (-215:Assertion failed) (unsigned)(i1 * DataType<_Tp>::channels) < (unsigned)(size.p[1] * channels()) in function 'at'


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/alalek/projects/opencv/contrib/modules/viz/misc/python/test/test_simple.py", line 83, in test_viz_show_simple_widgets
    viz.showWidget("grid1", cv.viz_PyWGrid((7,7), (0.75,0.75), cv.viz_PyColor().gray()), cv.viz_PyAffine3().translate((0.0, 0.0, -1.0)))
SystemError: <class 'cv2.viz_PyWGrid'> returned a result with an error set

======================================================================
ERROR: test_viz_show_trajectories (test_simple.viz_test)
----------------------------------------------------------------------
cv2.error: OpenCV(4.5.2-pre) /home/alalek/projects/opencv/dev/modules/core/include/opencv2/core/mat.inl.hpp:872: error: (-215:Assertion failed) (unsigned)(i1 * DataType<_Tp>::channels) < (unsigned)(size.p[1] * channels()) in function 'at'


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/alalek/projects/opencv/contrib/modules/viz/misc/python/test/test_simple.py", line 382, in test_viz_show_trajectories
    viz.showWidget("sub5", cv.viz_PyWTrajectoryFrustums(sub5, (0.78, 0.78), 0.15, cv.viz_PyColor().magenta())) #BUG
SystemError: <class 'cv2.viz_PyWTrajectoryFrustums'> returned a result with an error set

======================================================================
ERROR: test_viz_tutorial3_camera_view (test_simple.viz_test)
----------------------------------------------------------------------
cv2.error: OpenCV(4.5.2-pre) /home/alalek/projects/opencv/dev/modules/core/include/opencv2/core/mat.inl.hpp:872: error: (-215:Assertion failed) (unsigned)(i1 * DataType<_Tp>::channels) < (unsigned)(size.p[1] * channels()) in function 'at'


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/alalek/projects/opencv/contrib/modules/viz/misc/python/test/test_simple.py", line 52, in test_viz_tutorial3_camera_view
    tutorial3(True, self.find_file("viz/dragon.ply"))
  File "/home/alalek/projects/opencv/contrib/modules/viz/misc/python/test/test_simple.py", line 35, in tutorial3
    myWindow.showWidget("CPW_FRUSTUM", cv.viz_PyWCameraPosition((0.889484, 0.523599)), camera_pose)
SystemError: <class 'cv2.viz_PyWCameraPosition'> returned a result with an error set

======================================================================
ERROR: test_viz_tutorial3_global_view (test_simple.viz_test)
----------------------------------------------------------------------
cv2.error: OpenCV(4.5.2-pre) /home/alalek/projects/opencv/dev/modules/core/include/opencv2/core/mat.inl.hpp:872: error: (-215:Assertion failed) (unsigned)(i1 * DataType<_Tp>::channels) < (unsigned)(size.p[1] * channels()) in function 'at'


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/alalek/projects/opencv/contrib/modules/viz/misc/python/test/test_simple.py", line 49, in test_viz_tutorial3_global_view
    tutorial3(False, self.find_file("viz/dragon.ply"))
  File "/home/alalek/projects/opencv/contrib/modules/viz/misc/python/test/test_simple.py", line 35, in tutorial3
    myWindow.showWidget("CPW_FRUSTUM", cv.viz_PyWCameraPosition((0.889484, 0.523599)), camera_pose)
SystemError: <class 'cv2.viz_PyWCameraPosition'> returned a result with an error set

----------------------------------------------------------------------
Ran 22 tests in 30.685s

FAILED (errors=5)
```

File: modules/viz/include/opencv2/viz/widgets.hpp [hunk 14]
```diff
@@ -842,7 +930,833 @@ namespace cv
...
+                {
+                    Mat k = K.getMat();
+                    if (k.rows == 3 && k.cols == 3)
+                    {
+                        Matx33d x = k;
+                        widget = cv::makePtr<cv::viz::WCameraPosition>(x, image, scale, color.c);
+
+                    }
+                    else if (k.rows == 2 && k.cols == 1)
+                        widget = cv::makePtr<cv::viz::WCameraPosition>(Vec2d(k.at<double>(0, 0), k.at<double>(0, 1)), image, scale, color.c);
```
Comment (alalek) [inline modules/viz/include/opencv2/viz/widgets.hpp:1224]: There are several exceptions from Python tests here (in debug builds).

It is better to replace:
- check: `else if (k.total() == 2)`
- use `.at()` with single arg: `k.at<double>(0), k.at<double>(1)`

Causes by "undefined" 1D layout represented as 2D Mat (it may be row-based or column-based).

---

Please replace similar cases of `.at()` usage of 1D arrays below (use single argument form).

File: modules/viz/include/opencv2/viz/widgets.hpp [hunk 15]
```diff
@@ -842,7 +930,833 @@ namespace cv
...
+                    else
+                        CV_Error(-5, "unknown size");
+                }
+                else
+                    CV_Error(-5, "unknown type");
```
Comment (alalek) [inline modules/viz/include/opencv2/viz/widgets.hpp:1229]: > -5

Avoid using of magic numbers. Use enum values instead.

**Comment by LaurentBerger** [comment]
Hi,
About this [comment](https://github.com/opencv/opencv_contrib/pull/2882#pullrequestreview-613540731)
full output for my configuration 
[python config.txt](https://github.com/opencv/opencv_contrib/files/6155122/python.config.txt)

[python_test.txt](https://github.com/opencv/opencv_contrib/files/6155021/python_test.txt)

**Comment by alalek** [comment]
@LaurentBerger Please create branch in your "opencv" fork with commit from here: https://github.com/alalek/opencv/commits/python_fix_wrap_as
And then grab changes from here: https://github.com/alalek/opencv_contrib/commits/pr2882_r

**Comment by LaurentBerger** [comment]
Something is weird. I can see buildbot is ok with win64 but not on my configuration

 ```
1>------ Rebuild All started: Project: opencv_viz, Configuration: Release x64 ------
1>Building Custom Rule G:/Lib/opencv_contrib/modules/viz/CMakeLists.txt
1>opencv_viz_pch.cpp
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(662): error C2338: Matx should have at least 16 elements. channels >= 16
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(661): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(342): note: see reference to class template instantiation 'cv::Matx<double,4,1>' being compiled
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(1083): note: see reference to class template instantiation 'cv::Vec<double,4>' being compiled
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(650): error C2338: Matx should have at least 14 elements. channels >= 14
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(649): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(640): error C2338: Matx should have at least 12 elements. channels >= 12
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(639): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(629): error C2338: Matx should have at least 10 elements. channels >= 10
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(628): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(619): error C2338: Matx should have at least 9 elements. channels >= 9
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(618): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(610): error C2338: Matx should have at least 8 elements. channels >= 8
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(609): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(601): error C2338: Matx should have at least 7 elements. channels >= 7
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(600): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(592): error C2338: Matx should have at least 6 elements. channels >= 6
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(591): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(584): error C2338: Matx should have at least 5 elements. channels >= 5
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(583): note: while compiling class template member function 'cv::Matx<double,4,1>::Matx(_Tp,_Tp,_Tp,_Tp,_Tp)'
1>        with
1>        [
1>            _Tp=double
1>        ]
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(1091): error C2338: for arbitrary-size vector there is no cross-product defined cn == 3
1>G:\Lib\opencv\modules\core\include\opencv2/core/matx.hpp(1090): note: while compiling class template member function 'cv::Vec<double,4> cv::Vec<double,4>::cross(const cv::Vec<double,4> &) const'
1>Done building project "opencv_viz.vcxproj" -- FAILED.
========== Rebuild All: 0 succeeded, 1 failed, 0 skipped ==========
```

My opencv repo is even with master and I deleted build opencv repo

**Comment by alalek** [comment]
> with win64

Buildbot's Win64 configuration doesn't have VTK.
Only Linux has installed VTK dependency (but didn't run tests).

Do you have upgraded main "opencv" repository?

File: modules/viz/include/opencv2/viz/types.hpp [hunk 16]
```diff
@@ -60,64 +60,71 @@ namespace cv
 
         /** @brief This class represents color in BGR order.
         */
-        class Color : public Scalar
+        class CV_EXPORTS_W_SIMPLE Color
+#ifndef OPENCV_BINDING_PARSER
+            : public Scalar
+#endif
```
Comment (alalek) [inline modules/viz/include/opencv2/viz/types.hpp:66]: /cc @mshabunin Perhaps we need some `CV_WRAP_WITHOUT_BASE` marker for bindings.

**Comment by LaurentBerger** [comment]
> Do you have upgraded main "opencv" repository?

my [opencv repo](https://github.com/LaurentBerger/opencv) is now 2 commits behin master. Is it ok?

**Comment by alalek** [comment]
It should be OK.

Do you observe error?
- from your commit 5e5a2d47664a29ed8ce8c0a830aab5fb6b4654f8
- from new commit 8bf772cb36d34619dbc5947a3a15ef3ca0063e1a (please re-run CMake, because new files are added)
- without PCH headers: `cmake ... -DENABLE_PRECOMPILED_HEADERS=OFF`

**Comment by LaurentBerger** [comment]
without VTK viz is off no error

**Comment by alalek** [comment]
@LaurentBerger I'm able to reproduce this build problem. Investigating...

**Review by alalek — APPROVED** [review]
Well done! Thank you :+1:
